### PR TITLE
Explain what caused a given node to be built

### PR DESCRIPTION
**Pull request by richard-fine**
A common question we will have when working on builds is "why did thing X get built?" Today tunda will build a node if it is new, or if the hash of its inputs (action, input files, etc) has changed, but just knowing "the hash changed" isn't very helpful when it is a hash of 1000 things.

This PR makes three major changes:

1. It changes the way node GUIDs are computed to use the list of outputs rather than the list of inputs. This means that when a node's inputs are changed, we will still be able to retrieve the BuildState for it, and compare the new inputs to the saved information about previous inputs.

2. It adds a 'hash component log' which records the name/role of each individual component being fed to the hash, and saves it into the buildstate for that node. When the overall hashed input signature of the node changes, we can then compare the individual entries in the log to work out exactly which aspect of the input is different.

3. It adds a 'structured' log file, and reports new and changed nodes through it. We do this instead of just using stdout to make it easier to build other tools on top, as the format ('streamed' JSON, i.e. one JSON object per line) is easy for machines to parse, and we don't have to worry much about human-readability this way.

### CODE REVIEW

File: src/Hash.hpp [hunk 1]
```diff
@@ -152,13 +153,45 @@ struct ALIGN(16) HashStateImpl
...
+
 struct ALIGN(16) HashState
 {
   HashStateImpl m_StateImpl;
   uint64_t      m_MsgSize;
   size_t        m_BufUsed;
   uint8_t       m_Buffer[64];
   void*         m_DebugFile;
+  
+  HashComponentLog* m_ComponentLog;
```
Comment (lucasmeijer) [inline src/Hash.hpp]: let's get all this stuff out of the hashstate struct, and instead use LoggedHash functions that take both a hash and a structuredlog*

File: src/NodeState.hpp [hunk 2]
```diff
@@ -47,6 +47,8 @@ struct NodeState
   const char**              m_ImplicitDeps;
 
   HashDigest                m_InputSignature;
+  uint32_t                  m_FirstInputSignatureComponentIndex;
```
Comment (lucasmeijer) [inline src/NodeState.hpp]: let's turn these into a struct with index & length

File: src/Hash.cpp [hunk 3]
```diff
@@ -60,14 +97,13 @@ void HashSingleString(HashDigest* digest_out, const char* string)
...
-  while (true)
-  {
-    char c = *path++;
-    if (c == 0)
-      return;
-      c = FoldCase(c);
-      HashUpdate(self, &c, 1);
-  }
+  const size_t strLength = strlen(path);
+
```
Comment (lucasmeijer) [inline src/Hash.cpp]: maybe use a fixed buffer first, and only if it's not big enough use alloca with the strlen()

File: src/Hash.cpp [hunk 4]
```diff
@@ -47,6 +49,41 @@ void HashUpdate(HashState* self, const void *data_in, size_t size)
   
   self->m_BufUsed  = used;
   self->m_MsgSize += size * 8;
+
+  if (self->m_ComponentLog != nullptr && self->m_NextComponentName != nullptr)
+  {
```
Comment (lucasmeijer) [inline src/Hash.cpp]: extract to function

**Review by lucasmeijer — CHANGES_REQUESTED** [review]

File: src/NodeState.hpp (hunk 2 above)
Comment (richard-fine) [reply src/NodeState.hpp]: 6999845017817d534cbd75821fb3fee703e0d912

File: src/Hash.cpp (hunk 3 above)
Comment (richard-fine) [reply src/Hash.cpp]: e5eca5c22d7668a1b96a2fdcee5e4dc54f4039d3

File: src/Hash.cpp (hunk 4 above)
Comment (richard-fine) [reply src/Hash.cpp]: e5eca5c22d7668a1b96a2fdcee5e4dc54f4039d3

File: src/Hash.hpp (hunk 1 above)
Comment (richard-fine) [reply src/Hash.hpp]: e5eca5c22d7668a1b96a2fdcee5e4dc54f4039d3

**Pull request closed without merging**